### ISSUE TITLE
Level 4 of Lab 6

### ISSUE THREAD

**Issue by hengyjj** [nus-cs2030/2122-s2#115]
# Summary

Description of the question / problem here.
Hello! I am currently stucked at Level 4 and I am not too sure if using orElse is the correct way. This is my code.

<img width="771" alt="image" src="https://user-images.githubusercontent.com/77519191/159151599-32f7e842-7ab0-4c64-ad65-16ba56ee1e62.png">

So first I would check on whether a student is inside the map, and if it isn't, I would run the orElse(...). However, I am not too sure how to insert a new Student into the roster. I couldn't use the put() method as it returns a roster. Can anyone please advise me on how to attempt this?

Thank you!

Clarifications or details with respect to the issue you are facing

## Code Snippets

Attach code snippets (if any)

## Screenshots

Screenshots (if any)

**Comment by michael192002** [comment]
Hi, I was also stuck for hours at level 4, and I finally managed to finish it by using orElse(T t). I believe that get() in Optional is prohibited by CodeCrunch, so you might need to chain functions with orElse(T t) to return the value inside the Optional.

**Comment by hengyjj** [comment]
> 

Hi Michael. Thanks for your reply. I am not too sure what do you mean by chain functions? Currently, I am unclear as to how to add the student inside the ImMap if it does not exist using the orElse function.

**Comment by LauWyeRock** [comment]
It's the pipeline that prof talked about. Like flatMap(...).orElse()
The examples shown in level 3 are also examples of chaining functions

**Comment by michael192002** [comment]
For me, I did an implementation with a similar idea, but I had a getMap() method to retrieve the ImmutableMap, then I chained the get(stuId) with the orElse(...) like you did. I used stored the result of the Student object, and used it to retrieve the subclass (Module) if it exists, and so on. In the end, I did a series of put() calls based on the stored object variables.

**Comment by hengyjj** [comment]
> It's the pipeline that prof talked about. Like flatMap(...).orElse() The examples shown in level 3 are also examples of chaining functions

I see. Thanks. When I tried the chain functions, it keep giving me this error. 

<img width="771" alt="image" src="https://user-images.githubusercontent.com/77519191/159152930-b8d15d07-89a8-4bd6-ab7f-a256292617e4.png">

<img width="841" alt="image" src="https://user-images.githubusercontent.com/77519191/159152939-6cd6403d-ed5e-4b85-ad66-e2a4c199f8ee.png">

**Comment by michael192002** [comment]
The error resulted because the orElse method returns a Module, not an Optional<Module>. It seems that you are assigning the return value of orElse to Optional<Module>

**Comment by hengyjj** [comment]
> For me, I did an implementation with a similar idea, but I had a getMap() method to retrieve the ImmutableMap, then I chained the get(stuId) with the orElse(...) like you did. I used stored the result of the Student object, and used it to retrieve the subclass (Module) if it exists, and so on. In the end, I did a series of put() calls based on the stored object variables (my implementation is not the best, as it does not use flatMap())

I see. So I should first check if the student exists, and if it doesn't, I will use the getMap() to add the new student inside the map? Similar for Module as well?

**Comment by wk1267** [comment]
in addition, I think your 2nd line should be Student student = student.orElse(...).
I feel like .orElse() is basically just a .get() except with more functionality

**Comment by michael192002** [comment]
Sorry, turns out the getMap() was unnecessary; 

> > For me, I did an implementation with a similar idea, but I had a getMap() method to retrieve the ImmutableMap, then I chained the get(stuId) with the orElse(...) like you did. I used stored the result of the Student object, and used it to retrieve the subclass (Module) if it exists, and so on. In the end, I did a series of put() calls based on the stored object variables (my implementation is not the best, as it does not use flatMap())
> 
> I see. So I should first check if the student exists, and if it doesn't, I will use the getMap() to add the new student inside the map? Similar for Module as well?

Sorry, turns out the getMap() was unnecessary; all I did was just retrieve a Student (if found) and create a new Student object if not, and stored it in a variable, using it to invoke the method in the subclass Module. However, take note that for the Assessment part, for my implementation there is a special case

**Comment by hengyjj** [comment]
> Sorry, turns out the getMap() was unnecessary;
> 
> > > For me, I did an implementation with a similar idea, but I had a getMap() method to retrieve the ImmutableMap, then I chained the get(stuId) with the orElse(...) like you did. I used stored the result of the Student object, and used it to retrieve the subclass (Module) if it exists, and so on. In the end, I did a series of put() calls based on the stored object variables (my implementation is not the best, as it does not use flatMap())
> > 
> > 
> > I see. So I should first check if the student exists, and if it doesn't, I will use the getMap() to add the new student inside the map? Similar for Module as well?
> 
> Sorry, turns out the getMap() was unnecessary; all I did was just retrieve a Student (if found) and create a new Student object if not, and stored in a variable (either retrieved or new Student), using it to invoke other methods. However, take note that for the Assessment part, for my implementation there is a special case. In the end, I incorporated the stored variables to the roster, updating it. (Again, I think that my implementation of add() is not the best because it does not use mapping functions, so if you want, you can consider implementing with mapping functions)

Thanks Michael. May I know how did you retrieve the student from the Optional<Student> if student exists?

**Comment by michael192002** [comment]
I used the orElse(T t) method, because it returns the value inside the Optional if present, and T t if not

**Comment by hengyjj** [comment]
> I used the orElse(T t) method, because it returns the value inside the Optional if present, and T t if not

ahh so what u meant was like this? Thank you.

```
Optional<Student> student = this.roster.get(studID)

Student stud = student.orElse(new Student(studID));
```

**Comment by michael192002** [comment]
Yup, I guess my implementation has that logic (somewhat), all the best on your lab!

**Comment by hengyjj** [comment]
Thank you guys for all your help! Will keep this thread here for anyone who needs it too. :)
